### ISSUE TITLE
Momoization of Stream.tail()

### ISSUE THREAD

**Issue by danieldietrich** [rocketscience-projects/javaslang#74]
The lazy [javaslang.collection.Stream](https://github.com/rocketscience-projects/javaslang/blob/master/src/main/java/javaslang/collection/Stream.java) has intermediate computation results. The head of a Stream is computed immediately, the computation of the tail is deferred until needed.

**Note:** In contrast to java.util.stream.Stream, a javaslang.collection.Stream does not need an extra terminal operation like collect() for further processing of its elements.

```java
Stream.of(1, 2, 3)
head: 1, tail: () -> Stream.of(2, 3)

Stream.of(1, 2, 3).map(i -> i * 2) // = 2, 4, 6
head: 2, tail: () -> Stream.of(2, 3).map(i -> i * 2)

Stream.of(1, 2, 3).map(i -> i * 2).map(i -> i - 1) // = 1, 3, 5
head: 1, tail: () -> Stream.of(2, 3).map(i -> i * 2).map(i -> i - 1)
```

Multiple tail() calls lead to repeatedly evaluating both map applications when computing the head:

```java
Stream.of(1, 2, 3).map(i -> i * 2).map(i -> i - 1).tail()
head: 3, tail: () -> Stream.of(3).map(i -> i * 2).map(i -> i - 1)
```

The goal of this enhancement is to memoize tail in order to evaluate it only once.
